svgnative/ports/cairo/CairoSVGRenderer.cpp

```
@@ -0,0 +1,437 @@
+    for (int i = 0; i < path->num_data; i += path->data[i].header.length )
+    {
+        pathTransformed->data[i].header = path->data[i].header;
+        for (int j = 1; j < path->data[i].header.length; j ++ )
+        {
+            double x = path->data[i+j].point.x;
+            double y = path->data[i+j].point.y;
+            cairo_matrix_transform_point( &matrix, &x, &y );
+            pathTransformed->data[i+j].point.x = x;
+            pathTransformed->data[i+j].point.y = y;
+        }
+    }
+    cairo_path_destroy( path );
+    return pathTransformed;
+}
+
+void CairoSVGRenderer::Save(const GraphicStyle& graphicStyle)
+{
+    SVG_ASSERT( mCairo );
+    cairo_save ( mCairo );
+
+    if (graphicStyle.transform)
+        cairo_transform( mCairo, &(static_cast<CairoSVGTransform*>(graphicStyle.transform.get())->mMatrix));
+
+    if (graphicStyle.clippingPath && graphicStyle.clippingPath->path)
+    {
+        cairo_path_t* path = getTransformedClippingPath( graphicStyle.clippingPath.get() );
+        cairo_append_path( mCairo, path );
+        cairo_clip( mCairo );
+        cairo_path_destroy( path );
+        // FIXME: Cairo has no API to control winding/even-odd clipping. See
+        // https://lists.cairographics.org/archives/cairo/2009-September/018104.html
+    }
+}
+
+void CairoSVGRenderer::Restore()
+{
+    SVG_ASSERT( mCairo );
+    cairo_restore( mCairo );
+}
+
+inline void createCairoPattern(const Paint& paint, float opacity, cairo_pattern_t** pat)
+{
+    *pat = NULL;
+
+    if (paint.type() != typeid(Gradient))
+        return;
+
+    const auto& gradient = boost::get<Gradient>(paint);
+
+    SVG_ASSERT( gradient.type <= GradientType::kRadialGradient );
+
+    // in Cairo, gradient type might be set before setting color stops. See
+    // https://www.cairographics.org/samples/gradient/
+    switch (gradient.type)
+    {
+    case GradientType::kLinearGradient:
```

No default: case?

---

OK, I would add throw() for undefined gradient type.